File: 1140-stone-game-ii/1140-stone-game-ii.cs
public class Solution {
    private int n;
    private int[ , ] dp;
    private int[] sums;
    private int[] piles;
    
    public int StoneGameII(int[] piles) {
        this.n = piles.Length;
        this.dp = new int[n, n];
        this.sums = new int[n];
        this.piles = piles;
        
        sums[n-1] = piles[n-1];
        for(int i = n-2; i >= 0; i--){
            sums[i] = sums[i + 1] + piles[i];
        }
        
        return helper(0, 1);
    }
    
    public int helper(int i, int m){
        
        
        if(i + m * 2 >= n){
            return this.sums[i];
        }
        
        if(this.dp[i, m] != 0){
            return this.dp[i, m];
        }
        
        int min = int.MaxValue;
        
        for(int x = 1; x <= 2 * m; x++){
            if(i + x < n){
                 min = Math.Min(min, helper(i+x, Math.Max(x, m)));
            }
        }
        
        this.dp[i, m] = this.sums[i] - min;
        return this.dp[i, m];
    }
}

File: 1140-stone-game-ii/1140-stone-game-ii.java
class Solution {
    public int stoneGameII(int[] piles) {
        int n = piles.length;
        
        int[] sums = new int[n];
        sums[n-1] = piles[n - 1];
        
        // sum[i]: sum(piles[i:]) inclusively
        for(int i = n - 2; i >= 0; i--){
            sums[i] = sums[i + 1] + piles[i];
        }        
        int[][]dp = new int[n][n];
        
        return helper(0, 1, dp, sums, n);
        
        
    }
    
    private int helper(int i, int m, int[][] dp, int[] sums, int n){
        
        // not 'i + 2m >= n - 1' as i-th is the actual start of 2m
        if(i + 2 * m >= n){
            return sums[i];
        }
        
        if(dp[i][m] != 0){
            return dp[i][m];
        }
        
        int min = Integer.MAX_VALUE;
        
        // k <= 2m && i + k <= n - 1
        for(int k = 1; k <= Math.min(2 * m, n - i - 1); k++){
            min = Math.min(min, helper(i + k, Math.max(k, m), dp, sums, n));
        }
        
        dp[i][m] = sums[i] - min;
        
        return dp[i][m];
    }
}